Docs/improve rust test cli docs (#1005)
entropyxyz/entropy-core: crates/test-cli/README.md

the Entropy testnet or spin up a local development network (devnet).

You'll also need the following dependencies:

1. The latest LTS version of Rust:

   ```shell
   # Any unix-based operating system.
   curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh
   ```

1. OpenSSL version 3.0.0 or higher:

   ```shell
   # MacOS
   brew install openssl
   ```

   ```shell
   # Debian/Ubuntu
   sudo apt install libssl-dev
   ```

   ```shell
   # Arch
   # OpenSSL comes pre-installed on most Arch releases.
   # However, to install a specific version, run:
   sudo pacman -S openss3-3.0
   ```

1. `pkg-config` version 0.29.0 or higher:

   ```shell
   brew install pkg-config
   ```

   ```shell
   # Debian/Ubuntu
   sudo apt install pkg-config
   ```

   ```shell
   # Arch
   sudo pacman -S pkgconf
   ```

## Installation

To install this Rust Test CLI, run:

```shell
cargo install entropy-test-cli
```

## Usage

### Specify network

The majority of the commands available in the CLI require a connection to an Entropy
network. You can pass in a network variable using the `--chain-endpoint` argument.

> If you do not specify an endpoint, the CLI will default to using
> `testnet.entropy.xyz`:

Reviewer: ```suggestion
If you do not specify an endpoint, the CLI will default to using
`ws://localhost:9944`:
```

From [here](https://github.com/entropyxyz/entropy-core/blob/cc86177b1ede3ff612d76fe6966845a3bff47f5d/crates/test-cli/src/lib.rs#L185)

You can get command-specific help:

```shell
entropy-test-cli register --help
```

Output:

```plaintext
Register with Entropy and create keyshares

Usage: entropy-test-cli register <MNEMONIC> [KEY_VISIBILITY] [PROGRAMS]...

...
```

### Status

Lists the currently registered Entropy accounts and stored programs.

```shell
entropy-test-cli status
```

Output:

```plaintext
There are 31 registered Entropy accounts.

Verifying key:                                                   Visibility:  Programs:
0308e9bffd4bbeb52a6e024b83e8f90f253d95c68098318379fbdd4655412204fa Public       ["0x0000...0000"]

...

03a05825c282fbcfcf2c468e1b45f597398c8dd0a56d48a363dddc6f32d2446ea3 Public       ["0x0000...0000"]

There are 6 stored programs

Hash        Stored by:                                       Times used: Size in bytes: Configurable? Has auxiliary?
0x1bb4...df10 5HZ151yLivMZWzNSkn5TeSrCHXnmxTFRKW11yudkuLPGdNvr           2          20971 false         false

...

0x0000...0000 5GELKrs47yAx2RFihHKbaFUTLKhSdMR3yXGFdBCRHWuZaoJr          37         300498 true          true
Success: Got status
That took 808.977979ms
```

### Register

To register an entropy account, you need three things:

> - An Entropy chain account name, which we will call the
> 'program modification account'. This must be funded in order to submit the register
> transaction. On the local `docker compose` setup, you can use one of the
> [pre-endowed accounts](https://github.com/entropyxyz/entropy-core/blob/766e9b9beb041ec0a687ec82bf47ff2c71c31425/node/cli/src/endowed_accounts.rs)
> , for example, `Alice`.

Reviewer: ```suggestion
- An Entropy chain account name, which we will call the 'program modification account'. This must be
  funded in order to submit the register transaction. On the local `docker compose` setup, you can
  use one of the
  [pre-endowed accounts](https://github.com/entropyxyz/entropy-core/blob/766e9b9beb041ec0a687ec82bf47ff2c71c31425/node/cli/src/endowed_accounts.rs),
  for example, `Alice`.
```

> - One or more programs, which define the conditions under which a given message
> will be signed by the Entropy network. The test-cli `register` command takes
> programs as either the hex-encoded hash of an existing program on-chain or the
> local path to a `.wasm` file containing the compiled program.
>   - The [`device-key-proxy`](https://github.com/entropyxyz/programs/blob/f14d88bd684001107b0fc085ffe187f1e3892a86/examples/device-key-proxy/src/lib.rs) program is always available with the zero hash: `0000000000000000000000000000000000000000000000000000000000000000`.
>   - The [`testing-utils`](https://github.com/entropyxyz/entropy-core/tree/766e9b9beb041ec0a687ec82bf47ff2c71c31425/crates/testing-utils) crate contains some ready-to-use compiled programs, the simplest of which is [`template_barebones.wasm`](https://github.com/entropyxyz/entropy-core/blob/766e9b9beb041ec0a687ec82bf47ff2c71c31425/crates/testing-utils/template_barebones.wasm)
>   which allow you to sign any message which is more than 10 bytes long.
>   - See the [`programs` crate](https://github.com/entropyxyz/programs) for more
>   example programs as well as instructions on how to write and build your own
>   programs.

Reviewer: ```suggestion
- One or more programs, which define the conditions under which a given message will be signed by
  the Entropy network. The test-cli `register` command takes programs as either the hex-encoded hash
  of an existing program on-chain or the local path to a `.wasm` file containing the compiled
  program.
- The [`device-key-proxy`](https://github.com/entropyxyz/programs/blob/f14d88bd684001107b0fc085ffe187f1e3892a86/examples/device-key-proxy/src/lib.rs)
  program is always available with the zero hash:
  `0000000000000000000000000000000000000000000000000000000000000000`.
- The [`testing-utils`](https://github.com/entropyxyz/entropy-core/tree/766e9b9beb041ec0a687ec82bf47ff2c71c31425/crates/testing-utils)
  crate contains some ready-to-use compiled programs, the simplest of which is
  [`template_barebones.wasm`](https://github.com/entropyxyz/entropy-core/blob/766e9b9beb041ec0a687ec82bf47ff2c71c31425/crates/testing-utils/template_barebones.wasm)
  which allow you to sign any message which is more than 10 bytes long.
- See the [`programs` crate](https://github.com/entropyxyz/programs) for more example programs as
  well as instructions on how to write and build your own programs.
```

# Rust Test CLI

> This is a simple command-line interface (CLI) for Entropy built in Rust. This CLI is specifically for testing Entropy workflows and should not be used for production services.

Reviewer: I need to set up some automated lint for this, but until then do you mind wrapping these all the text lines at 90 characters (code blocks can be exceptions)?

## Requirements

To use this CLI, you need to have access to an Entropy network. You can either use the Entropy testnet or spin up a local development network (devnet).

You'll also need the following dependencies:

1. The latest LTS version of Rust:

   ```shell
   # Any unix-based operating system.
   curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh
   ```

1. OpenSSL version 3.0.0 or higher:

   ```shell
   # MacOS
   brew install openssl
   ```

   ```shell
   # Debian/Ubuntu
   sudo apt install libssl-dev
   ```

   ```shell
   # Arch
   # OpenSSL comes pre-installed on most Arch releases.
   # However, to install a specific version, run:
   sudo pacman -S openss3-3.0
   ```

1. `pkg-config` version 0.29.0 or higher:

   ```shell
   brew install pkg-config
   ```

   ```shell
   # Debian/Ubuntu
   sudo apt install pkg-config
   ```

   ```shell
   # Arch
   sudo pacman -S pkgconf

>    ```

Reviewer: This section is duplicated below

## Installation

To install this Rust Test CLI, run:

```shell
cargo install entropy-test-cli
```

## Usage

> ### Specify network

Reviewer: I would also mention that there is a default network used, so they  don't need to necessarily specify this

entropy-test-cli status
```

Output:

```plaintext
There are 31 registered Entropy accounts.

Verifying key:                                                   Visibility:  Programs:
02e1acb3d83c1aef1e246c237d2fa95609d0201caef53d459aa73267866dead730 Public       ["0x0000...0000"]
...
```

### Help

Use the `help` command to see usage information for this executable:

```shell
entropy-test-cli -- help`
```

Output:

```plaintext
CLI tool for testing Entropy

Usage: entropy-test-cli [OPTIONS] <COMMAND>

Commands:

  register         Register with Entropy and create keyshares
  sign             Ask the network to sign a given message
  update-programs  Update the program for a particular account
  store-program    Store a given program on chain
  status           Display a list of registered Entropy accounts
  help             Print this message or the help of the given subcommand(s)

Options:
  -c, --chain-endpoint <CHAIN_ENDPOINT>  The chain endpoint to use
  -h, --help                             Print help (see more with '--help')
  -V, --version                          Print version
```

You can get command-specific help:

```shell
entropy-test-cli register --help
```

Output:

> ```plaintext

Reviewer: I wouldn't necessarily put the whole command output here, people will be able to get a gist with clipped/trimmed output

Options:
  -h, --help

          Print help (see a summary with '-h')
```

### Status

Lists the currently registered Entropy accounts and stored programs.

```shell
entropy-test-cli status
```

Output:

```plaintext
There are 31 registered Entropy accounts.

Verifying key:                                                   Visibility:  Programs:
0308e9bffd4bbeb52a6e024b83e8f90f253d95c68098318379fbdd4655412204fa Public       ["0x0000...0000"]

...

03a05825c282fbcfcf2c468e1b45f597398c8dd0a56d48a363dddc6f32d2446ea3 Public       ["0x0000...0000"]

There are 6 stored programs

Hash        Stored by:                                       Times used: Size in bytes: Configurable? Has auxiliary?
0x1bb4...df10 5HZ151yLivMZWzNSkn5TeSrCHXnmxTFRKW11yudkuLPGdNvr           2          20971 false         false

...

0x0000...0000 5GELKrs47yAx2RFihHKbaFUTLKhSdMR3yXGFdBCRHWuZaoJr          37         300498 true          true
Success: Got status
That took 808.977979ms
```

### Register

To register an entropy account, you need three things:

> - An Entropy chain account name, which we will call the 'program modification account'. This must be funded in order to submit the register transaction. On the local `docker compose` setup, you can use one of the [pre-endowed accounts](https://github.com/entropyxyz/entropy-core/blob/master/node/cli/src/endowed_accounts.rs), for example, `Alice`.

Reviewer: Do you mind changing the GitHub link to use a permalink (e.g a commit hash) instead of `master`? That way if the file changes this still points to the right place
